src/compiler/compiler.rs

```diff
@@ -275,124 +296,205 @@ impl Compiler {
+        let creator = creator.clone();
+
+        Box::new(result.and_then(move |preprocessor_result| {
+            // If the preprocessor failed, just return that result.
+            if !preprocessor_result.status.success() {
+                debug!("[{}]: preprocessor returned error status {:?}",
+                       parsed_args.output_file(),
+                       preprocessor_result.status.code());
+                // Drop the stdout since it's the preprocessor output, just hand back stderr and the exit status.
+                let output = process::Output {
+                    stdout: vec!(),
+                    ..preprocessor_result
+                };
+                return Box::new(future::ok((CompileResult::Error, output))) as Box<Future<Item=_, Error=_>>
+            }
+            trace!("[{}]: Preprocessor output is {} bytes",
+                   parsed_args.output_file(),
+                   preprocessor_result.stdout.len());
+
+            let key = hash_key(&me, &arguments, &preprocessor_result.stdout);
+            trace!("[{}]: Hash key: {}", parsed_args.output_file(), key);
+            // If `ForceRecache` is enabled, we won't check the cache.
+            let start = Instant::now();
+            let cache_status = if cache_control == CacheControl::ForceRecache {
+                Box::new(future::ok(Cache::Recache))
+            } else {
+                storage.get(&key)
+            };
+
+            Box::new(cache_status.then(move |result| {
+                let duration = start.elapsed();
+                let pwd = Path::new(&cwd);
+                let outputs = parsed_args.outputs.iter()
+                    .map(|(key, path)| (key.to_string(), pwd.join(path)))
+                    .collect::<HashMap<_, _>>();
+
+                let miss_type = match result {
+                    Ok(Cache::Hit(mut entry)) => {
+                        debug!("[{}]: Cache hit!", parsed_args.output_file());
+                        let mut stdout = io::Cursor::new(vec!());
+                        let mut stderr = io::Cursor::new(vec!());
+                        entry.get_object("stdout", &mut stdout).unwrap_or(());
+                        entry.get_object("stderr", &mut stderr).unwrap_or(());
+                        let write = pool.spawn_fn(move ||{
+                            for (key, path) in &outputs {
+                                let mut f = try!(File::create(path));
+                                try!(entry.get_object(&key, &mut f));
+                            }
+                            Ok(())
+                        });
+                        let output = process::Output {
+                            status: exit_status(0),
+                            stdout: stdout.into_inner(),
+                            stderr: stderr.into_inner(),
+                        };
+                        let result = CompileResult::CacheHit(duration);
+                        return Box::new(write.map(|_| {
```

Comment:
Heh yeah it took me awhile to get the hang of them but I've found them to be super useful  once you get the hang